### ISSUE TITLE
Input/Output Errors and PCI devices unavailable after suspend

### ISSUE THREAD

**Issue by danjjeff** [QubesOS/qubes-issues#3049]
#### Qubes OS version (e.g., `R3.2`):
3.2 and 4.0rc1

#### Affected TemplateVMs (e.g., `fedora-23`, if applicable):
dom0, sys-net

---

### Expected behavior:
Qubes can be suspended and recover from suspend

### Actual behavior:
After suspend Qubes is unstable.
Behavior is inconsistent. Sometimes networking is just disabled and nmcli in the sys-net system VM reports that the ethernet and wireless devices are unavailable and system is otherwise fine. At other times the sys-net VM is unresponsive or shuts down completely and dom0 gives input/output errors when attempting to open terminals or shutdown. The errors from dom0 are also bizarre as they affect different command from one test to the next. Sometimes lspci will throw the error, other times dmesg, ls or less will throw an error and lspci is fine. Often initctl will give the input/output error and the system must be restarted.

### Steps to reproduce the behavior:
Suspend qubes (close lid, use menu and echo mem > /sys/power/state all produce the same result)
Awaken (lift lid, push power button, it ignores keystrokes on my laptop)
### General notes:
Hardware is a Lenovo X1 Carbon gen3, wireless adapter is Intel 7265 rev 59. I've run full system diagnostics on the laptop and it passes. I've tested different nvme drives with no benefit.

I've tried the steps from  https://github.com/QubesOS/qubes-issues/issues/2922 and https://www.qubes-os.org/doc/wireless-troubleshooting/#automatically-reloading-drivers-on-suspendresume, but they have not helped.

---

#### Related issues:
https://groups.google.com/forum/#!topic/qubes-users/LkP-6ORGwME
 https://github.com/QubesOS/qubes-issues/issues/2922

**Comment by andrewdavidwong** [comment]
This sounds like it might be a duplicate of #3008. The workaround is to blacklist `iwlmvm`. (See issue comments for details.)

**Comment by andrewdavidwong** [comment]
(If it turns out not to be a duplicate, let me know, and we can reopen this.)

**Comment by danjjeff** [comment]
I wish it were. That is in the troubleshooting steps recommended in the wireless-troubleshooting doc. Unfortunately adding iwlmvm to /rw/config/suspend-module-blacklist made made no difference for me in testing.

**Comment by danjjeff** [comment]
I'm going to test downgrading to 3.2 and rolling back the kernel to see if that corrects it.

**Comment by z4ppy** [comment]
Hello,

I had the same issue on 3.2, it seems resolved since the last update for me
(4.9.35-19.pvops.qubes.x86_64)

2017-08-24 20:29 GMT+02:00 Daniel Jeffery <notifications@github.com>:

**Comment by danjjeff** [comment]
@z4ppy, unfortunately I didn't have it _until_ I updated to 4.9.35-19 :(

**Comment by rtiangha** [comment]
@danjeff: Yes, you need to restart sys-net once you've modified the blacklist file. You also need to make sure to blacklist both iwlmvm and iwlwifi in the file, not just iwlmvm.

**Comment by danjjeff** [comment]
I blacklisted both iwlmvm and iwlwifi. I restarted Qubes entirely. The problem persisted. Just to be sure I am not remembering incorrectly or making some mistake, I will try it again with the latest kernel.

I have reinstalled 3.2 fresh with it running 4.4.14-11 and there is no problem. Suspend works fine and I do not get the odd lockups or input/output errors. I'll update now to latest 4.9.35-15 and put in the blacklist lines.

**Comment by danjjeff** [comment]
It looks like I spoke too soon. The sys-net VM had not crashed after the suspend on the older kernel and NetworkManager reported the wifi connection was still up, but it wasn't passing any traffic and once I downed the connection nmcli couldn't bring it up again.

I've blacklisted iwlwifi and iwlmvm in sys-net:/rw/config/suspend-module-blacklist and restarted the sys-net VM. The behavior from the previous paragraph was exactly repeated.

A key difference worth noting in the behavior on the older kernel is that Network Manager still thinks the connection is active and the device is connected. Also, I don't seem to be getting the bizarre behavior in dom0.

**Comment by danjjeff** [comment]
Okay, on 3.2 with kernel 4.9.35-19 and the fedora 25 template, I am currently seeing the same suspend behavior as on 4.4.14-11 with fedora 23. sys-net:/rw/config/suspend-module-blacklist contains iwlmvm and iwlwifi, each on their own line.

Since I don't seem to be getting the input/output errors anymore (for no reason?) on 3.2 I guess I'll stay here for now and just not suspend. I am very open to other ideas or troubleshooting.

**Comment by danjjeff** [comment]
Hooked back up my USB mouse and found the sys-net VM has the same problems. After suspend, USB is also broken until the VM is restarted.

And now the input/output errors are back in dom0 and I can't stop and restart the VMs :)

I'm not sure if that is the result of just running it long enough or because I restarted the VMs and suspended a second time without a reboot, but they're back. I'm really wondering if this is a hardware failure at this point, but all the Lenovo system diagnostics come back fine.

**Comment by rtiangha** [comment]
I don't know; these symptoms are weird. I have an Intel 7260 dual ac card and it seems to work fine, although it's not a 7265 but one would think it was close enough.

I noticed in your log output on the mail list that it couldn't load the wifi firmware. Just to double check, but is it actually installed (I assume it is, but you never know)?

sudo dnf install iwl7260-firmware or sudo dnf install linux-firmware (I'm not sure which; I'm a Debian guy)

Also, check the Lenovo website for any BIOS updates and if they exist, try applying them. Maybe this is a known hardware issue that's already been fixed and there are a few cases out there with similar symptoms on other distros and they all seem to come from Lenovo users so maybe this is something the manufacturer has already addressed in a BIOS update.

**Comment by danjjeff** [comment]
It is weird. BIOS is a good point and I had upgraded the BIOS to latest right at the beginning of troubleshooting. The iwl7260 firmware is installed correctly by default. I don't know that the wireless or the USB layers are the right place to look at this point, though. This seems to be an issue with PCI passthrough after suspend or some other events that happen with time since I've had the issue start just after the machine has been running for a while.

The most frustrating part of this is the inconsistency. I have not had this problem since updating to 3.2 as soon as it released last year, but now it's present even on fresh 3.2 install. The problem seems to have started about 2 weeks ago.

I was able to get a hold of another identical (checked all the hardware chips and revisions) gen 3 X1 Carbon and compare against 3 more identical machines running Qubes, but not in my possession. The two in my possession that I have wiped and tested on Qubes 4, 3.2 as-installed and 3.2 up-to-date all exhibit the same behavior. Two of the other 3 seem to also exhibit the suspend behavior, but not the freezes while the 3rd is reported to be fine and is fully patched like the others. 

I went ahead and booted up a live disk of Kali and there seem to be no issues. Suspend works fine and there are no unusual freezes or input/output errors. I'm at a bit of loss where to even look next, but at this point I can only use Qubes if I don't let it suspend and even then, it has repeatedly locked up on me and lost work in progress. 

If it will help, I am perfectly willing to overnight or 2-day one of these laptops to a dev to help sort this out.

**Comment by rtiangha** [comment]
As a last resort, maybe try one of @fepitre's 4.12 kernels that he posted to the mail list to see if it helps? The kernel options shouldn't be much different than 4.9's except for the new drivers introduced, but maybe the power management stuff is better. It seems it's kind of flakey in 4.9, especially when it comes to Intel wifi; having Intel power management disabled in the kernel makes suspend work for some cards and not others, and enabling it in the kernel flips it around (currently, it's disabled in-kernel because having it enabled was causing too many issues, but there's a sysctl or kernel value you can toggle to enable it yourself, but I don't know it off the top of my head).

https://sourceforge.net/projects/qubes-linux-kernel/files/

**Comment by danjjeff** [comment]
Well, I thought maybe I'd gotten somewhere over the weekend as I reinstalled 3.2 again and ... everything worked. I suspended and restarted several times and everything seemed fine, so I crossed my fingers and updated the kernel and the template VM, but then it went back to locked up PCI devices and input/output errors. I tried setting everything to use the old kernel, 4.4.14-11, but the problems persisted. I tried Reg's suggestion of using the 4.12 kernels and still had the same problems.

My conclusion at this point is that while the kernel may be involved, it's not just the kernel that's the problem. I'm going to try reinstalling 3.2 fresh, _again_, and see if I can get the state where everything works. I have done a fresh install of 3.2 about 5 times in the last several days and only had it work correctly that once, so I'm not optimistic it will work, but I am at a loss as to why it would be different. BIOS settings are the same, install parameters are the same.

One issue I don't think I've mentioned is that when this was working correctly, the system can shutdown and restart successfully. When it's in the bad configuration it will hang on shutdown. In the bad configuration before anything appears to be broken (network and usb still working fine, no input/output errors) if I attempt to shutdown or restart I can watch it attempting 8 different background processes, all of which appear to be dismounts until it hits the 1m30s limit and then hang with a series of errors like `device-mapper: remove ioctl on [device] failed: Device or resource busy`. At this point I'm forced to hard power off by holding down the power button.

Once I've done a suspend or waited long enough for input/output errors and/or net and usb errors, when I attempt to restart I get `blk_update_request: I/O error, dev sda sector [some number that changes each time]`. The second error made me wonder about bad nvme, but swapping it didn't change anything.

**Comment by rtiangha** [comment]
So just to confirm, installing fresh from ISO works, updating system afterwards doesn't, and switching to an older kernel from that point still doesn't?

If you're going to re-install from scratch, can you a) capture dmesg output from *both* dom0 and sys-net and/or attach system logs when it's working, then *just* update kernel, kernel-qubes-vm (and kernel-devel if you have it) in dom0 and try it again with the new kernel and report back if it still works or not (capture dmesg if it doesn't)? And if it does work, update dom0 and sys-net's template with the regular system updates and try again?

**Comment by rtiangha** [comment]
Also, at each step, verify the running kernel in both dom0 and sys-net by running `uname -r` as a sanity check.

**Comment by danjjeff** [comment]
I changed the title to better reflect this does not appear to be primarily about the network. Both USB and Network VMs lose their devices and dom0 is having issues even running dmesg and sometimes lspci or reading logs.

@rtiangha It doesn't always work fresh from the ISO. I reinstalled 3.2 at least 5 times over the last week and only once did it work correctly. It's reinstalling right now. If it works, I'll collect the logs, update just the kernel and kernel-qubes-vm packages on dom0 and see what that gets us. As noted, I probably can't capture dmesg when it's not working as that command throws the input/output error nearly all the time once we're in the bad state (as well as trying to less/vi/grep anything in /var/log). I've been uname'ing for exactly that reason all along the way. :) Thanks for your help.

**Comment by rtiangha** [comment]
Cool. Keep the post updated. Full logs where possible would be helpful to at least see what's going on. Personally, I've never seen this behaviour ever.

Also, am I correct in thinking that you've got sys-net acting as a combined USBvm as well, or do you have a separate sys-usb VM?

**Comment by danjjeff** [comment]
For all the tests over the last week I've had separate sys-usb and sys-net VMs. Are there any logs other than dmesg you'd like me to capture?

**Comment by danjjeff** [comment]
The thing driving me nuts about this behavior is how inconsistent it's being. I'm trying to hold my install and config parameters totally consistent and think of anything I or the hardware are doing that could give different results, but I'm at a bit of a loss. One difference I just thought of between the two machines I'm testing with and the other three also running Qubes is the BIOS revision. These two are fully patched and I'm not sure if the other three have ever been patched from what the factory shipped.

**Comment by rtiangha** [comment]
Any kind of system logs would be useful. Do it in dom0, sys-net, and sys-usb. We're still in fact-finding mode.

Also, what kind of BIOS options does the machine have when it comes to Virtualization? On the surface, this looks like something buggy with VT-d.

Finally, you say you have a set of 2 machines with an updated BIOS and a set of 3 that doesn't? When this stuff works, on what set of machines does it actually work on?

**Comment by danjjeff** [comment]
In BIOS there is an option for Intel (R) Virtualization Technology and Intel (R) VT-d Feature. Both are enabled.

Of the 3 other machines, 1 seems to be fine and having no issues. One seems to be exhibiting the same issues as the two in my possession on suspend, but doesn't seem to develop problems simply by being on for some period of time. The third is having occasionally freezes that require a hard power off (this is the behavior I was seeing before the network lockups and input/output errors started), but the third machine seems fine recovering from the suspend. I've confirmed that all three of those machines are running the same BIOS revision.

**Comment by danjjeff** [comment]
I'm having the problem immediately on the fresh install. I did the exact same fresh install twice on Friday. The first time was like this, with the problem behavior occurring immediately, the second time it was working fine until I updated the kernel. Today's fresh install is having the problems immediately.

Fortunately, this time it let me run dmesg while sys-net and sys-usb were unable to reach the hardware dedicated to them. I've attached the tarred up dmesg files.


[dmesg-logs-bad-config-fresh-3.2.tar.gz](https://github.com/QubesOS/qubes-issues/files/1257461/dmesg-logs-bad-config-fresh-3.2.tar.gz)

**Comment by danjjeff** [comment]
Reinstalled 3.2 two more times on the same hardware and it has had the broken suspend right out of the install both times.

**Comment by rtiangha** [comment]
So are you saying that 1 machine out of the 5 works no matter what you do to it, and the others don't? What BIOS version are all of these running?

**Comment by danjjeff** [comment]
/sigh

I was attempting to just use one of the two systems I've been reinstalling and not suspend. So, I wrote a full response in firefox on the personal VM, then Qubes froze and I lost it. So, starting again.

All are identical gen 3 Lenovo X1 Carbons purchased as a single batch and running Qubes since late 2015.

In use by others:
1 of 5: BIOS 1.10, identical BIOS settings to my broken boxes. Qubes 3.2. Fedora 25 template VMs. No issues.
2 of 5: BIOS 1.10, BIOS settings not verified. Qubes 3.2. Fedora 25 template VMs. net/usb pci devices unavailable in VMs after a suspend (usually, works fine ~10% of the time).
3 of 5: BIOS 1.10, BIOS settings not verified. Qubes 3.2. Fedora 25 template VMs. Occasional OS freezes (every 3-4 days?). No suspend issues.

In my possession:
4 of 5: BIOS 1.17, identical BIOS settings to 1 of 5. Qubes 3.2 (also tested 4). Fedora 23 or 25 template VMs. PCI devices unavailable in VMs after suspend. If actually running firefox in one VM, it freezes after about 30 minutes. Reinstalled 4 twice and 3.2 3 times in the last week.
5 of 5: BIOS 1.17, identical BIOS settings to 1 of 5. Qubes 3.2 (also tested 4). Fedora 23 or 25 template VMs. PCI devices unavailable in VMs after suspend. If actually running firefox in one VM, it freezes after about 30 minutes. Installed and tested 4, reinstalled fresh 3.2 5 times in the last 4 days, worked fine on one install until kernel was updated.

Looking at this, I'll try rolling the BIOS back to 1.10. The problems 2 and 3 are having are not as consistent or comprehensive as the problems 4 and 5 are having. I know there are some rollback prevention BIOS settings I'll probably need to play with.

**Comment by rtiangha** [comment]
Or rather than rolling back to 1.10, figure out what the BIOS settings of Number 3 is and copy it to the machines that aren't working. If Number 3 is the one that's working most of the time, you need to figure out what makes it different.

**Comment by danjjeff** [comment]
1 of 5 is the one with no issues. 3 has freezing issues fairly regularly. 4 and 5 have their BIOS settings set identical to 1's.

**Comment by rtiangha** [comment]
But your message said "3 of 5: BIOS 1.10, BIOS settings not verified. Qubes 3.2. Fedora 25 template VMs. Occasional OS freezes (every 3-4 days?). No suspend issues."???

**Comment by rtiangha** [comment]
Edit: Oops, my bad. I misread.

**Comment by danjjeff** [comment]
Correct. 3 is not having issuing coming out of suspend, but it freezes up completely roughly every 72 hours.

**Comment by danjjeff** [comment]
Ugh, actually I realized that is incomplete. 3 of 5 is running primarily Debian user VMs, instead of Fedora. I believe the system VMs are updated to Fedora 25, but the user VMs were Debian 8 and they upgraded to 9 to see if it would fix things, but it didn't help.

**Comment by rtiangha** [comment]
Well, maybe downgrading to 1.10 might help. There's also a 1.18 BIOS update that was released on Aug 22. Might be worth trying?

**Comment by danjjeff** [comment]
Okay, tentative progress, maybe, if it sticks, I hope. I went to download the 1.10... hehe. You found it too :)

I just installed 1.18 on 5 of 5 and restarted. Did not kill PCI coming out of suspend. I did nothing else to the box and it was reliably locking up immediately before this. So, probably a very good sign. Doing some more testing to see if this continues, then I'll update the kernel and see what happens.

**Comment by rtiangha** [comment]
Well, I was reading through changelogs and it looks like you won't easily be able to downgrade past 1.14. But maybe Lenovo is aware of the issues and hopefully 1.18+ will work better.

**Comment by danjjeff** [comment]
I was afraid that might be the case based on some of the BIOS settings I'd been reviewing.

So, I'm seeing I need to control the method used to put the machine into suspend and test variations. For the last several days I have pretty consistently been using the method described in another ticket of `echo mem > /sys/power/state` in dom0. This still produces an unstable system even after updating to BIOS 1.18. 

However, if I use the menu > logout > suspend route or close the lid, system 4 of 5 (currently fresh 3.2, 4.4.11-14 kernel, fedora 23 templates) suspends and recovers just fine. I tested this suspend route on 5 of 5 (currently running 4.9.35-19 kernel, all other dom0 updates applied and fedora 25 templates) and it failed to recover from suspend at all, simply freezing up the window manager and returning a blank screen, although other tty sessions (ctrl+alt+f2..f6) were accessible. I updated to BIOS 1.18 and the behavior remained the same.

At this point I've updated ONLY kernel, kernel-qubes-vm and perl-math-bigint (dependency) on 4 of 5 and rebooted. Immediately, recovering from suspend broke the wifi, the device was unavailable in the sys-net vm, but USB was fine. So I added iwlwifi and iwlmvm to /rw/config/suspend-module-blacklist and restarted the sys-net vm but this did not help. On returning from suspend after this change I began to receive the block I/O errors and a lot of read only file system errors as dm-1 had been reloaded ro.

I rebooted and tried again with the suspend-module-blacklist in place and now 4 of 5 is behaving like 5 of 5 and displaying only an unresponsive black screen instead of the xscreensaver login prompt. So, from working to broken, by updating _only_ the kernel. I'm going to try setting EFI to use the older kernel on 4 of 5 and 5 of 5 and see what happens then.

**Comment by danjjeff** [comment]
And that seems to nail it. On 4 of 5 nothing but the kernel was updated and the system chokes on suspend, but reverting the kernel fixes the issue. On 5 of 5 everything was updated and all VM templates are Fedora 25. It chokes after suspend, but reverting to using the 4.4.14-11 kernel allows suspend to work fine, so the problem is definitely tied to the new kernel.

Again, this is suspending by using the app menu, not by `echo mem > /sys/power/state` which seems to always cause a problem.

**Comment by danjjeff** [comment]
I tried the 4.12.8-20 kernel again on 5 and it gets stuck in endless reboot.

So, we seem to have an interaction between BIOS and the post-4.4 kernels. Latest BIOS from Lenovo makes it possible to use the 4.4 kernel without the major problems I was seeing on 4.9, but I'm now running an older, unpatched kernel.

**Comment by rtiangha** [comment]
And what about the 4.9 kernel?

**Comment by danjjeff** [comment]
The previous two comments to the one on 4.12 were about comparing the 4.9 and 4.4 kernels. It will boot on 4.9, but consistently and reliably breaks after suspend and freezes after about a half hour of using the system. The breaking on suspend results in the I/O errors, inability to restart itself, some filesystems going read only and PCI devices being unavailable to the VMs they were assigned to.

**Comment by danjjeff** [comment]
BTW, Reg, thanks for jumping in and helping me troubleshoot this. It's very appreciated.

**Comment by rtiangha** [comment]
Well, I'm going to have to dig deep into the kernel config options between 4.4 and 4.9 to figure out if there's anything different.

In the meantime, you can try to compile the 4.4 kernel off of my branch here:

https://github.com/rtiangha/qubes-linux-kernel/tree/stable-4.4

That will get you the latest 4.4 version that was released last week. If the problem *still* persists, it might be an upstream problem.

**Comment by rtiangha** [comment]
Also, just curious: Does at least blacklisting the wifi modules work properly in BIOS 1.18?

**Comment by rtiangha** [comment]
Finally, just shooting from the hip, but did you run ME cleaner on your Thinkpad?

**Comment by rtiangha** [comment]
Also, looks like the 4.9.45 kernel is slowly making its way to current-testing. I'm doubtful it'll solve things in this case, but you never know. But yeah, if you can compile a newer 4.4 kernel and report back, it'll help with the troubleshooting. In my head, there might be a few more things to try, but it'll require changing some kernel options to narrow things down.

**Comment by danjjeff** [comment]
Interesting thought, but, no, ME cleaner has not been run. In BIOS there is an option to permanently disable AMT and that has been done. I'll give compiling the newer 4.4 kernel a shot. 

I'm not sure how to define 'works properly' where using suspend-module-blacklist is concerned. I have put the entries in, but can't see any change in the system performance regardless of whether they are present or not. I was considering also adding cfg80211 and mac80211 to see if that made it possible for the wireless to recover. That said, I really don't think that's where the problem lies as none of these systems have needed that set in the past and the USB becoming unavailable, plus the Input/Output errors and dom0 filesystems going read only seem to point to a problem lower level and more pervasive than a simple suspend/resume power management issue with iwl.

**Comment by rtiangha** [comment]
By "works properly" I meant if the wifi works after suspend. There could be different issues at play here.

It's the "30 minutes" that interests me. If it's always 30 minutes before it breaks, it might be related to the ME (but again, it's a shot in the dark). The newer kernel versions have the  ME driver compiled out of it (for obvious reasons, including the fact that AMT makes no sense in a Qubes context), but I wonder if for machines like yours, it's actually needed, or if machines were provisioned for AMT in the past (some manufacturers do that before shipping), maybe the driver is needed as well. That's why if you can compile your own kernels, it'd help with the testing since we can toggle on and off various options to see what works.

**Comment by danjjeff** [comment]
I got 4.4.84-14 compiled and installed and it seems to be working well so far.

**Comment by danjjeff** [comment]
I tried compiling 4.9, but ran into some errors so I'm setting up to compile on a FC23 backed build VM (I compiled the 4.4.84-14 kernel on a fc25 backed VM). Once I'm set up I'll go through and review the options and see if I can see anything that seems like it would matter. Any suggestions you have on what to look at first would be welcome.

**Comment by rtiangha** [comment]
Try the standard 4.9.45 first. Who knows, maybe this is an upstream thing after all.

**Comment by danjjeff** [comment]
K. I just finished reading through the diff of the 4.4 and 4.9 config files. There are a lot of changes to consider.

**Comment by rtiangha** [comment]
Not surprising considering all the new things introduced between 4.4 and 4.9. That's 5 major revisions. Does 4.9.45 work as is?

**Comment by danjjeff** [comment]
Been testing it. I had to figure out a couple of not-very-obvious dependencies that 4.9 needed for compiling and I've been playing with different EFI settings. I don't if you're familiar with the EFI troubles for booting with the Lenovo BIOS, but you have to set mapbs=1 and noexitboot=1 in xen.cfg. 

With 4.4 this sorted things out, but now with 4.9 I get these messages from efi immediately before the disk encryption password prompt: 
```efi: EFI_MEMMAP is not enabled.
efi: Failed to lookup EFI memory descriptor for 0x00000000a3dbc018```

So, no it doesn't appear to be fixed.

**Comment by rtiangha** [comment]
So what happens then? Does it continue to boot or does it hang? I don't have a working UEFI system (I have to use legacy boot) so I don't know what the expected behavior is.

4.9.46 was just released last night. If compiling, all you need to do is change the number in the 'version' and recompile.

**Comment by danjjeff** [comment]
Sorry, that last comment wasn't very clear. 4.9.45 will boot, but throws those two errors. It would seem these should be resolved by mapbs. Right now I'm pulling the xen source to find a listing of the xen.cfg parameters related to efi since I couldn't find them documented anywhere, but mailing lists.

I've tried several different things in BIOS and xen.cfg changes using 4.9.45-21 and gotten very frustrating, inconsistent results. I've rebooted about 30-40 times since my last post and I can hold all the parameters the same and get different behaviors, apparently at random.

Typically, the system boots up and I open Firefox in the work VM, browse a single page and then use the app menu to logout > suspend. Even before suspending, there are inconsistencies as the laptop is a QHD display and Firefox opens at one of 3 different HiDPI compensation levels. Most commonly it's insanely large, 42 point printed font everything and barely usable, sometimes it's at a reasonable DPI and other times it opens at screen native and the text is quite small. If it opens at either of the latter two DPIs it has never successfully suspended in the following steps. The DPI issues I know how to deal with and had it managed in the past, but I'm leaving it vanilla here for testing and the inconsistency from one boot to the next, when nothing has been changed, is super fun.

At this point one of two things happens, either I briefly see the unlock menu then the screen goes blank (backlight still on), then fully suspends or it just goes straight to blank and then shuts off the screen. 

On pushing the power button to resume, it just powers on the backlight and either completely ignores all keyboard/touchpad input, appearing frozen, or pulls up the unlock dialog on the first keypress and lets me log in.

Now I either have a fully working system (I'd say this happened about 5 times) or I can see network manager showing it's disconnected and the system will be sluggish and close the firefox window after a few minutes. Attempting to sudo in the dom0 terminal I keep open will get input/output errors and attempting to shut down will lock up with the blkread errors listed above.

At this point I'm inclined to think that the inexplicable successful 'installs' were really just inexplicable successful boots. This kind of inconsistency screams 'race condition' to me, but might also be related to EFI and memory management. I'm going to look some more at the xen efi mess and see what I can come up with, but any other suggestions would be welcome.

@andrewdavidwong @marmarek have you seen anything like this in your testing, particularly the possibility of this being related to efi?

**Comment by danjjeff** [comment]
Failed to suspend on the first boot with the new 4.9.46 kernel. One aspect i didn't mention is that even when it does suspend successfully, all but one time it would fail on the the second attempt to suspend without a reboot. This time around, it's suspended twice without dying, but I am jaded at this point and expect to be betrayed on the next boot.

**Comment by danjjeff** [comment]
Lost networking, USB and started getting I/O errors after the third suspend. I got 4 suspends in on the next boot after I checked /proc/acpi/wakeup. It appears most of the devices are kernel blacklisted so I added SLPB and LID. Still locked up after the 4th suspend. I am also looking into /etc/mkinitcpio.conf MODULES="intel_agp i915" to see if that fixes the instances of the blank screen that won't respond to the keyboard.

I've been searching in the kernel, xen and qubes-linux-kernel repos and I can't find the noexitboot parameter listed anywhere, which is odd. mapbs seems like the only applicable xen parameter. I'm going to play with some efi parameters for the kernel itself and see if they help. Possibly try setting reboot=acpi.

**Comment by danjjeff** [comment]
I went ahead and installed via Legacy boot (UEFI totally disabled) on 4 of 5 to see if that helped. It didn't. 4.9.35 kernel still loses net and usb and sends several mounts read only.

**Comment by andrewdavidwong** [comment]
> have you seen anything like this in your testing, particularly the possibility of this being related to efi?

I'm sorry, but I haven't seen anything like this before. :frowning_face:

**Comment by danjjeff** [comment]
Well, I'm backing off on this for now. At this point what we know is that the up-to-date BIOS (1.18) for the Lenovo X1 Carbon Gen 3 is not entirely stable with Qubes/Xen Kernels later than 4.4. UEFI does not seem to be the problem as the instability is still present on a straight legacy boot with UEFI completely bypassed. 

The intermittent system freezes (usually about 1/2 hour after system startup) seem to be resolved on kernel 4.9.46. I've been running a system for over 24 hours with a few VMs open and some traffic with no spontaneous freezes yet.

The issues with loop mounts going read-only and PCI passthrough devices (USB and network) becoming unavailable in the VM are definitely still present, though they do seem to manifest somewhat less frequently. Using the suspend-module-blacklist did not seem to help.

**Comment by rtiangha** [comment]
More shooting from the hip: Is Wake-On-LAN enabled in the BIOS? If so, try disabling it and see if it helps with the wifi issues.

Also, you mention "Input/Output" errors. Where do these appear? Neither the issue description nor the Related Issues link actually has an example of those errors.

**Comment by danjjeff** [comment]
Wake-On-LAN has always been set to off. Thanks for circling back with a suggestion.

Regarding the I/O error messages, I haven't put in the exact line. Sorry. It's really simple though and IMHO is just BASH saying that things that shouldn't be read only are read only. To reproduce, once the system has recovered from suspend, but sys-net and sys-usb report the passed-through devices are unavailable (many times they both just become unresponsive and will even close a terminal session I had open prior to suspend), type something like sudo dmesg in the dom0 terminal and you will get: `standard.input: Input/Output Error`

At this point, the system will not allow me to retrieve logs or otherwise interact with it. Attempting to reboot will lead to the block device errors (seems to be loop devices that have gone read only) and reboot/shutdown will typically lock up.

I have made some progress with the X1 Carbon Gen5. Since it's kabylake we've got to have a newer kernel so I've rolled my own Qubes 4 build with fepitre's 4.12 kernel and gotten it installed and running in legacy mode. I've given minimal effort to attempting to get that running in EFI mode, but all the current range of tricks I'm familiar with (and I think I know most of the ones that have worked with Qubes in the past at this point) did not get it to boot via EFI.

Currently, I'm trying to return to getting work done on the Gen3 (5 of 5 in this ticket) and as long as I don't let it suspend it's been pretty stable on 4.9.46. I did finally have it freeze up completely on me yesterday and today while it was running three VMs, all of which were active, but not particularly pushing the memory or cpu. I may just roll this one back the 4.8 kernel and see if it's more stable. I have not yet had success booting it with the 4.12 kernel. Other than changing the kernel it's a 'vanilla' Qubes 3.2 install still.

**Comment by rtiangha** [comment]
One other thing to try (another one from the hip) is that if the latest 4.4 kernel works (as of yesterday, it's at 4.4.87), then to take that config file and migrate it to a 4.9 config. You can do this by copying the 4.4 config file and overwrite the 4.9 config file in the 4.9 repo branch. Run "make prep" to extract the kernel source (as of yesterday, latest is 4.9.48), copy that file to kernel-4.9.xx/linux-4.9.xx/.config, enter the directory and run 'make oldconfig.' It'll then prompt you on what to do with every new kernel option that was introduced in between 4.4 and 4.9 (I think you can probably just stick with the defaults for now, or if you really want, tweak to your needs). Once done and saved, copy back that new .config file over the config file in the root directory and compile that. If the problems go away, then it's probably fixable by a missing kernel config option. If the problems persist, then it might be an upstream thing or a Qubes/4.9+ problem. But at least we might be able to narrow things down a bit more this way.

**Comment by danjjeff** [comment]
I got the 4.12 kernel/4.0 Qubes built on the Lenovo X1 Carbon Gen5 and once I had the template VMs setting up right, it lost the network and USB after a suspend. So, I'm sad there.

I tried what you suggested Reg, although I was worried as I went through it. There are a LOT of new kernel parameters (maybe 50?) and I just accepted defaults. That said, Qubes 3.2 on the Gen 3 rebooted and seems to run with the new kernel, but it is exhibiting the same lock up the network and USB, put systems read-only behavior after a suspend. I guess that tells us it's probably not a parameter that we changed from 4.4 to 4.9. Maybe it's one Qubes didn't change?

**Comment by rtiangha** [comment]
Are you absolutely certain that the original 4.4 config that you migrated to 4.9 worked properly in the beginning? Because if so, I have to admit, I'm a little stumped, although also relieved since it would indicate that it probably isn't due to a kernel config option that was enabled in 4.4 but is now missing in 4.9.

Yeah, as you noticed, there were a lot of options added in those 5 major kernel releases between the two, and it's a lot to dig through. If we wanted to be systematic, and assuming that 4.4 config worked properly on that machine so that it could be used as a baseline, the next thing to try would be to migrate that 4.4 config to a 4.5 config and see if that works properly, and if so, back it up and migrate that 4.5 config to 4.6, then 4.7, etc and then see where it stops working (final version numbers in each branch can be found [here](https://en.wikipedia.org/wiki/Linux_kernel#4.x.y_releases)). That would at least help narrow things down and maybe give some hints on what to look at to see what really changed between the release where it worked and the immediate release after where it stopped.

**Comment by danjjeff** [comment]
I'll run 4.4 again and verify it is having no issues, but when I tested it above, it seemed to run perfectly.

I'm hitting the limit of what I can keep pouring into this time-wise as I need to work and I'm now having to setup something of my own that gives me as much of what Qubes offers as I can quickly manage. Gotta get work done. I'll see if I can do the kernel progression you're suggesting on the side over the next few days, though.

**Comment by danjjeff** [comment]
I have tried to make the 4.4.87 kernel fail, but it's been rock steady under heavy load and repeated suspend resume cycles (about 8 in my test). I tried the 4.8 kernel building from your repo, rtiangha, and it failed suspend resume and put the system into i/o error with the r/o systems on the first suspend. I'll build a 4.7 kernel next.

**Comment by rtiangha** [comment]
Can you do me a favor and post the 4.4 config file that you're using as an attachment or a link? I just want to make sure we're working off the same files.

**Comment by dballard** [comment]
So I seem to be having a similar but opposite problem, that I've done less research into. 
When I boot my system76 laptop with an intel centrino wireless-n 1030 wifi chip, rfkill lists it as hard blocked. After suspending and resuming, it's suddenly not blocked and I can use it

```
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: Xen PCI mapped GSI18 to IRQ26
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: pci frontend enable msi failed for dev 0:0
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: Xen PCI frontend error: -22!
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: pci_enable_msi failed - -22
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: loaded firmware version 18.168.6.1 op_mode iwldvm
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: CONFIG_IWLWIFI_DEBUG enabled
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: CONFIG_IWLWIFI_DEBUGFS enabled
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: CONFIG_IWLWIFI_DEVICE_TRACING disabled
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: Detected Intel(R) Centrino(R) Wireless-N 1030 BGN, REV=0xB0
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0: RF_KILL bit toggled to disable radio.
Oct 27 07:56:09 sys-net kernel: iwlwifi 0000:00:00.0 wlp0s0: renamed from wlan0
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: Radio type=0x2-0x2-0x1
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:57:03 sys-net kernel: iwlwifi 0000:00:00.0: Radio type=0x2-0x2-0x1
```

Suspend/Resume

```
Oct 27 07:59:06 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:59:06 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:59:06 sys-net kernel: iwlwifi 0000:00:00.0: Radio type=0x2-0x2-0x1
Oct 27 07:59:07 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:59:07 sys-net kernel: iwlwifi 0000:00:00.0: L1 Enabled - LTR Disabled
Oct 27 07:59:07 sys-net kernel: iwlwifi 0000:00:00.0: Radio type=0x2-0x2-0x1
```

Let me know if there's anything i can do with my setup that might help you

**Comment by danjjeff** [comment]
@dballard What processor and chipset are you running?

I suspect that the problems are unrelated as I really don't believe the primary issue in this ticket (I think there is more than one problem at play here) is an issue with the wifi so much as an issue with the behavior after suspend. 

@rtiangha I missed your comment earlier. The config for 4.4 was totally unmodified from the default.

At this point I'm running the 4.9 kernel and never use suspend, always turning the laptop off completely when I will not be using it or commuting. While this has security benefits, it has been a significant drain on productivity and is a non-trivial irritation. I had hoped my days of running Linux in the 90s were behind me.

I have still had no luck running Qubes successfully on the X1 gen5, even using 4.10-12 kernels. I will probably try this again once 4.0 releases completely.

**Comment by marmarek** [comment]
> I have still had no luck running Qubes successfully on the X1 gen5, even using 4.10-12 kernels.

In about two weeks I might be able to debug this (i.e. have access to such machine).

**Comment by danjjeff** [comment]
Marek, if it's just a matter of access, I'd happily send you the one I have to test with.

**Comment by danjjeff** [comment]
@marmarek Any update on your success getting the suspend/resume working with a recent kernel or getting Qubes running on an X1 Carbon gen5?

**Comment by marmarek** [comment]
There were some delays (and a lot of work on other 4.0-related things), hopefully will do this or next week.

**Comment by danjjeff** [comment]
Thank you, @marmarek . The offer to send hardware to test on still stands.

**Comment by danjjeff** [comment]
Reviewing Qubes 4 open bugs I came across this, which sounds a lot like where I started at in this issue (as documented above, this issue is 100% present in Qubes 4 as well as 3.2): https://github.com/QubesOS/qubes-issues/issues/3359

**Comment by andrewdavidwong** [comment]
@marmarek, do you think #3359 is a duplicate of this?

**Comment by marmarek** [comment]
> @marmarek, do you think #3359 is a duplicate of this?

Not sure about it.

**Comment by marmarek** [comment]
> There were some delays (and a lot of work on other 4.0-related things), hopefully will do this or next week.

Ok, finally managed to install Qubes 4.0-rc3 on X1 Carbon 5th gen. As for network and suspend issues, for me helps:
 - update dom0 kernel to 4.14.6 (https://github.com/QubesOS/qubes-linux-kernel/pull/13)
 - keep VM kernel at 4.9.56-21
 - add `iwlmvm` and `iwlwifi` to `/rw/config/suspend-module-blacklist` in sys-net

But there are other issues, see here: https://groups.google.com/d/msg/qubes-users/zCLYDy1bRKw/ziwORQlMAAAJ
